test/core/bad_connection/close_fd_test.cc

```diff
@@ -0,0 +1,762 @@
+  op->data.recv_message.recv_message = &response_payload_recv;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
+  op->data.recv_status_on_client.trailing_metadata = &trailing_metadata_recv;
+  op->data.recv_status_on_client.status = &status;
+  op->data.recv_status_on_client.status_details = &details;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  error = grpc_call_start_batch(call, ops, static_cast<size_t>(op - ops),
+                                tag(1), nullptr);
+  GPR_ASSERT(GRPC_CALL_OK == error);
+
+  error = grpc_server_request_call(g_ctx.server, &server_call, &call_details,
+                                   &request_metadata_recv, g_ctx.bound_cq,
+                                   g_ctx.cq, tag(101));
+  GPR_ASSERT(GRPC_CALL_OK == error);
+  event = grpc_completion_queue_next(
+      g_ctx.cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(event.success == 1);
+  GPR_ASSERT(event.tag == tag(101));
+  GPR_ASSERT(event.type == GRPC_OP_COMPLETE);
+
+  memset(ops, 0, sizeof(ops));
+  op = ops;
+  op->op = GRPC_OP_SEND_INITIAL_METADATA;
+  op->data.send_initial_metadata.count = 0;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+  op->op = GRPC_OP_RECV_MESSAGE;
+  op->data.recv_message.recv_message = &request_payload_recv;
+  op->flags = 0;
+  op->reserved = nullptr;
+  op++;
+
+  grpc_endpoint_pair* sfd = g_ctx.ep;
+  int fd;
+  if (fdtype == SERVER_FD) {
+    fd = sfd->server->vtable->get_fd(sfd->server);
+  } else {
+    GPR_ASSERT(fdtype == CLIENT_FD);
+    fd = sfd->client->vtable->get_fd(sfd->client);
+  }
+  /* Connection is closed before the server receives the client's message. */
+  close(fd);
+
+  error = grpc_call_start_batch(server_call, ops, static_cast<size_t>(op - ops),
+                                tag(102), nullptr);
+  GPR_ASSERT(GRPC_CALL_OK == error);
+
+  event = grpc_completion_queue_next(
+      g_ctx.bound_cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+
+  /* Batch operation completes on the server side.
+   * event.success will be true if the op completes successfully.
+   * event.success will be false if the op completes with an error. This can
+   * happen due to a race with closing the fd resulting in pending writes
+   * failing due to stream closure.
+   * */
+  GPR_ASSERT(event.type == GRPC_OP_COMPLETE);
+  GPR_ASSERT(event.tag == tag(102));
+
+  event = grpc_completion_queue_next(
+      g_ctx.client_cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  /* When the client fd is closed, the server gets EPIPE.
+   * When server fd is closed, server gets EBADF.
+   * In both cases server sends GRPC_STATUS_UNAVALABLE to the client. However,
+   * the client may not receive this grpc_status as it's socket is being closed.
+   * If the client didn't get grpc_status from the server it will time out
+   * waiting on the completion queue
+   */
+  if (event.type == GRPC_QUEUE_TIMEOUT) {
```

Comment:
We should assert on the type of event here rather than an `if` at this place. If the type is TIMEOUT, for sure the asserts on `event.success == 0` and `event.tag == nullptr` are true, so these lines are like no-op. The correct event type depends on the expected behavior at this place.

Reply:
Talked with Muxi offline. When an fd is closed the server gets an error on write and it sends GRPC_STATUS_UNAVAILABLE to the client. The server side error looks like this:
```{"created":"@1549064622.947201486","description":"FD Shutdown","file":"src/core/lib/iomgr/lockfree_event.cc","file_line":194,"referenced_errors":[{"created":"@1549064622.947194884","description":"Delayed close due to in-progress write","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":587,"referenced_errors":[{"created":"@1549064622.947191634","description":"Bad file descriptor","errno":9,"fd":6,"file":"src/core/lib/iomgr/tcp_posix.cc","file_line":912,"grpc_status":14,"os_error":"Bad file descriptor","syscall":"sendmsg","target_address":"socketpair-server"}]}]}```

Since the fd is being closed, the client may or may not receive the server's response. That's why we have the `if` condition. The interesting thing here is that it looks like the client doing a read doesn't get notified that the socket has been closed.